lib/src/combine_latest.dart

```diff
@@ -0,0 +1,119 @@
+/// well, regardless of [combineWith]'s type. If a single subscription stream is
+/// merged into a broadcast stream it may never be canceled.
+StreamTransformer<S, R> combineLatest<S, T, R>(
+        Stream<T> combineWith, FutureOr<R> Function(S, T) combine) =>
+    _CombineLatest(combineWith, combine);
+
+class _CombineLatest<S, T, R> extends StreamTransformerBase<S, R> {
+  final Stream<T> _other;
+  final FutureOr<R> Function(S, T) _combine;
+
+  _CombineLatest(this._other, this._combine);
+
+  @override
+  Stream<R> bind(Stream<S> source) {
+    final controller = source.isBroadcast
+        ? StreamController<R>.broadcast(sync: true)
+        : StreamController<R>(sync: true);
+
+    final other = (source.isBroadcast && !_other.isBroadcast)
+        ? _other.asBroadcastStream()
+        : _other;
+
+    StreamSubscription sourceSubscription;
+    StreamSubscription otherSubscription;
+
+    var sourceDone = false;
+    var otherDone = false;
+
+    S latestSource;
+    T latestOther;
+
+    var sourceStarted = false;
+    var otherStarted = false;
+
+    void emitCombined() {
+      if (!sourceStarted || !otherStarted) return;
+      FutureOr<R> result;
+      try {
+        result = _combine(latestSource, latestOther);
+      } catch (e, s) {
+        controller.addError(e, s);
+        return;
+      }
+      if (result is Future<R>) {
+        result
+            .then(controller.add, onError: controller.addError)
+            .whenComplete(() {
+          sourceSubscription.resume();
+          otherSubscription.resume();
+        });
+      } else {
+        controller.add(result as R);
+      }
+    }
+
+    controller.onListen = () {
+      if (sourceSubscription != null) return;
+      sourceSubscription = source.listen(
+          (s) {
+            sourceStarted = true;
+            latestSource = s;
+            emitCombined();
+          },
+          onError: controller.addError,
+          onDone: () {
+            sourceDone = true;
+            if (otherDone) controller.close();
+          });
+      otherSubscription = other.listen(
+          (o) {
+            otherStarted = true;
+            latestOther = o;
+            emitCombined();
+          },
+          onError: controller.addError,
+          onDone: () {
+            otherDone = true;
+            if (sourceDone) controller.close();
+          });
+      if (!source.isBroadcast) {
+        controller
+          ..onPause = () {
+            sourceSubscription.pause();
+            otherSubscription.pause();
+          }
+          ..onResume = () {
+            sourceSubscription.resume();
+            otherSubscription.resume();
+          };
+      }
+      controller.onCancel = () async {
+        var cancelSource = sourceSubscription.cancel();
+        var cancelOther = otherSubscription.cancel();
+        sourceSubscription = null;
+        otherSubscription = null;
+
+        await cancelSource;
+        await cancelOther;
```

Comment:
Here you await two futures in order. If the second future completes with an error before the first one completes at all, that error will be uncaught.

Consider doing `await Future.wait([cancelSource, cancelOther]);`.

Reply:
Good catch. Thanks